src/nspawn/nspawn.c

```diff
@@ -2226,97 +2227,120 @@ static bool should_enable_fuse(void) {
 }
 
-static int copy_devnodes(const char *dest, bool enable_fuse) {
-        _cleanup_strv_free_ char **devnodes = NULL;
-        int r = 0;
+static int copy_devnode_one(const char *dest, const char *node, bool ignore_mknod_failure) {
+        int r;
 
         assert(dest);
+        assert(!isempty(node));
 
-        devnodes = strv_new("null",
-                            "zero",
-                            "full",
-                            "random",
-                            "urandom",
-                            "tty",
-                            STRV_IFNOTNULL(enable_fuse ? "fuse" : NULL),
-                            "net/tun");
-        if (!devnodes)
+        BLOCK_WITH_UMASK(0000);
+
+        _cleanup_free_ char *from = path_join("/dev/", node);
+        if (!from)
                 return log_oom();
 
-        BLOCK_WITH_UMASK(0000);
+        _cleanup_free_ char *to = path_join(dest, from);
+        if (!to)
+                return log_oom();
 
-        /* Create /dev/net, so that we can create /dev/net/tun in it */
-        if (userns_mkdir(dest, "/dev/net", 0755, 0, 0) < 0)
-                return log_error_errno(r, "Failed to create /dev/net directory: %m");
+        struct stat st;
+        if (stat(from, &st) < 0) {
+                if (errno != ENOENT)
+                        return log_error_errno(errno, "Failed to stat %s: %m", from);
 
-        STRV_FOREACH(d, devnodes) {
-                _cleanup_free_ char *from = NULL, *to = NULL;
-                struct stat st;
+                log_debug_errno(errno, "Device node %s does not exist, cannot copy it, ignoring.", from);
+                return 0;
+        }
+        if (!S_ISCHR(st.st_mode) && !S_ISBLK(st.st_mode))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                       "%s is not a char or block device, cannot copy.", from);
 
-                from = path_join("/dev/", *d);
-                if (!from)
-                        return log_oom();
+        /* Create the parent directory of the device node. Here, we assume that the path has at most one
+         * subdirectory under /dev/, e.g. /dev/net/tun. */
+        _cleanup_free_ char *parent = NULL;
+        r = path_extract_directory(from, &parent);
+        if (r >= 0) {
+                if (userns_mkdir(dest, parent, 0755, 0, 0) < 0)
+                        return log_error_errno(r, "Failed to create directory %s: %m", parent);
+        } else if (r != -EDESTADDRREQ)
+                return log_error_errno(r, "Failed to extract directory from %s: %m", from);
+
+        if (mknod(to, st.st_mode, st.st_rdev) < 0) {
+                /* Explicitly warn the user when /dev/ is already populated. */
+                if (errno == EEXIST)
+                        log_notice("%s/dev/ is pre-mounted and pre-populated. If a pre-mounted /dev/ is provided it needs to be an unpopulated file system.", dest);
+                if (!ERRNO_IS_PRIVILEGE(errno) || arg_uid_shift != 0) {
```

Comment:
Is this really the case where we want to ignore errors?

The report indicates that `errno` would be `EPERM`, which is a `ERRNO_IS_PRIVILEGED`, so the first alternative would not match. Rather all other sorts of `mknod` failures would be swallowed here. It still catches the reported case, because `EPERM` realistically only can happen when shifting uids.

I would have expected `if (ERRNO_IS_PRIVILEGE(errno) && arg_uid_shift != 0)` as condition here. Maybe you can add a comment explaining the choice of condition if it is meant as is?